README.md
# Nfsc

By Rudolf König, Jürgen Weiger *et al*.

## Intro

**Nfsc** is a terminal emulator. If this file is part of the p3nfs package,
then **Nfsc** is also the psion counterpart to the `p3nfsd` daemon on unix with
additional terminal emulation support.

For `p3nfsd` related info refer to the README file in the `p3nfsd` package.

**Nfsc** emulates a DEC VT100 terminal and some features of the newer vt220.
It has following noteworthy features:

- Support for all builtin Series 3a fonts (9 fonts on the S3a).
- Support for bold, underlined, inverse and italic fonts.
- Alternative character set support.
- Defineable function key support.
- Jumpscroll and charset conversion (ISO8859-1 <-> IBM codepage 850).
- Support for history scrollback (over 500 lines are possible).
- Bring server functionality.
- Paste from the scrollback or from builtin applications.
- Dialing/scripting support. Extensible, as it is done in OPL.
- More than one configuration can be saved/loaded.
- Online help.
- Support for baudrates 50 to 115200.
- Should work on Psion Series 3; it is reported to work on the Siena.
- Robust: most (all?) VT100 animations are working.
- X/YMODEM support (partially working).
- Local mode.
- Experimental 3Fax support.
- Automatic dial at startup.
- TTY:I (infrared).

The current release (5.4) fixes some bugs and adds some features mainly of
interest for non-casual *Nfsc* users. See `CHANGELOG.md` for more.

**License:** GNU General Public Licence (Version 2)

## Installation

Place `nfsc.app` in the `\app` directory and install it with **Psion-I**.
If you also want keypad support, place `fnkeys.nfs` into the `\opd`
directory. Read it (with a text editor) for details.

If you want dial support, install the files `dialme.opl` & `nfscdial.opl`
in the `opl` directory and translate them. See below for more.

The rest of the files are documentation and source code. They should not be
copied to the Psion.

## Building Nfsc (If you have the Psion C SDK)

In the `src` directory, from the DOS prompt:

```batch
C:\PROJECTS\NFSC> makeshd nfsc
C:\PROJECTS\NFSC> rcomp nfsc.hlp
C:\PROJECTS\NFSC> tsc /m nfsc
C:\PROJECTS\NFSC> ren nfsc.img nfsc.app
```

## Notes

- X/Ymodem was tested with the only Zmodem package I found for UNIX.
  Xmodem works, Ymodem has a problem: the receiver won't terminate cleanly
  after a file was sent. I think the Psion driver and the Unix program
  behave differently here.
- break. As there is no `p_break()` function or the like, the break is
  generated as a series of 0 bytes sent at 50 baud. This can work, but
  is not guaranteed, as the definition of break is 0 for at least 250ms
  and this method generates more 200ms pieces. It works on SunOS, but not
  on Solaris.
- Note for S3 owners: I only tested it on my S3a in compatibility mode.
  All but the online help should work, as it is too wide for the S3.
  Please drop me a mail if it works on your S3.
- Siena was reported to work OK.

## Dial Support

It is implemented in OPL, so if you want to use it, you have to write
first an OPL program, and have to translate it (typing Psion-T).

Don't be afraid, it may be sufficient to write a program as simple as:

```opl
PROC dialme:
  global nfscPid%
  loadm "nfscdial"
  dialinit:

  puts:("ATDT 0123456789"+chr$(10))
  expect:("login:")
  puts:("my_username"+chr$(10))
  expect:("Password:")
  puts:("my_password"+chr$(10))
ENDP
```

The function `hotk:("x")` calls **Nfsc** to execute "Psion-x".
You can of course add more expect/puts lines or use OPL features if you
like (or have to :)).

Dialing is a second "thread", it works in parallel, so you can type
on the psion while the script is listening (e.g. your password). The
only problem with this method is that the dial script may lose data.


### How Dial Support Works (for programmers only)

#### Sending text to Nfsc

The opl program sends a message with `ID $40` to the **Nfsc** application,
and the address of the OPL string.

#### Receiving data

The opl program sends a message with `ID $41` to the **Nfsc** application,
and waits for data. It sends a return receipt of 1 if it wants to
quit, 0 if it requests more data.

**Note:** if the OPL program receives a string of length 0, this is in fact
a modem status change report. The first data byte is to interpreted as
    follows:

```text
      Signal	Bit
      CTS	0
      DSR	1
      DCD	2
      RTS	3
      DTR	4
```

As it is a string of length 0 it won't disturb any normal OPL routine.

See `nfscdial.opl` for the implementation of `expect`, `puts` & `hotk`.
(It would be nice to have a "`gets`" call.)

## Tips

- You can save your modem parameters with a little opl "dialme" program
  like the following:

```opl
PROC dialme:
  global nfscPid%
  loadm "nfscdial"
  dialinit:

  puts:("ATZ"+chr$(10))
ENDP
```

- If you're logging in to a UNIX system via **Nfsc**, and you dont want to
  start the `p3nfsd` program there (the UNIX counterpart of **Nfsc**), you may
  have an inappropriate number of lines set. `p3nfsd` takes care of that,
  but you have to be root while starting it.

  Here is how to fix the number of rows (a little UNIX exercise):

  ```bash
  # Check the number of lines (rows) with the following command:
  % stty -a
  # To set the number of rows to 20, try
  % stty rows 20
  # If this won't work, then try
  % setenv LINES 20
  # If you now get "setenv: not found", then (you're using a different shell):
  $ LINES=20
  $ export LINES
  ```


- If you like to fiddle with the save files manually, then you can use the
  "include filename" statement. Include files can be nested and "Exit"
  will not overwrite the file if there was an "include" in it.

## History

**Nfsc** was originally developed as an OPL counterpart for the UNIX p3nfs
daemon which is a way to connect psions to UNIX computers.

After N.N. rewrote it as a C program, I (Rudi) added terminal emulator
code to it. Then Michael checked it once more together with me so that
the vt100 animations ran on it. Odd Gripensteim used it on a VMS system
and wished he had key translation, so he got it (but not everything he
wished for).

The famous WWW page ot Steve Litchfield helped me to reorganize the
whole, and add some user features to it, like online help and config
saved in a file. For a Bring server I have to remember the data on the
screen. From here was a little to get scrollback too.

## Sources

```text
nfsc.c        | protocol and dialog routines
vt100.c       | terminal emulator stuff + scrollback handling
xymodem.c     | X/Y modem support
paste.c       | bring + paste server
params.c      | parameter loading and saving
nfsc.hlp      | guess what
```


CHANGELOG.md
# nfsc 5.4 (Iain Tuddenham's wish list)

> Fri Jan 30 12:15:19 SGT 1998

- Now nfsc save files support the include directive:
  include \<path\> reads in the specified file. Note that
  in this case "exit" wont save the file.
- "Duplicate character in insert mode" bug fixed.
  This could be seen in programs like emacs and pine. They actually
  use insert mode.
- Some hotkeys changed to be more "psion" conform.
  (Exit without saving, Paste)
- Saving X/Ymodem mode bug fixed
- UNIX-Mode: Only the drives marked for exporting are exported to p3nfsd.
- Modem state changes are now reported to the OPL program.
- "Toggle echo" menu entry added for switching off the display, used for
  speeding up ascii downloads.
- Sending more than 127 bytes from the OPL program is handled correctly now.
- OPL program: lockups occured at sending data while waiting for text are
  eliminated. The data sent to nfsc will not be sent back to the OPL program,
  as the OPL program waits for the ack from the send, so it cannot listen
  for new data.
- limited the number of scrollback lines to 800 as the old limit (1000)
  caused allocating 81000-65536=15464 bytes


src/cnv.h
/*
  Twoway conversion between ISO8859-1 and cp850
  Only the upper 128 characters are affected
  Rudolf Koenig
 */

unsigned char tbl_iso2cp[128] = 
{
  0x9f, 0xb0, 0xb1, 0xb2, 0xb3, 0xb4, 0xb9, 0xba,
  0xbb, 0xbc, 0xbf, 0xc0, 0xc1, 0xc2, 0xc3, 0xc4,
  0xc5, 0xc8, 0xc9, 0xca, 0xcb, 0xcc, 0xcd, 0xce,
  0xd5, 0xd9, 0xda, 0xdb, 0xdc, 0xdf, 0xf2, 0xfe,
  0xff, 0xad, 0xbd, 0x9c, 0xcf, 0xbe, 0xdd, 0xf5,
  0xf9, 0xb8, 0xa6, 0xae, 0xaa, 0xf0, 0xa9, 0xee,
  0xf8, 0xf1, 0xfd, 0xfc, 0xef, 0xe6, 0xf4, 0xfa,
  0xf7, 0xfb, 0xa7, 0xaf, 0xac, 0xab, 0xf3, 0xa8,
  0xb7, 0xb5, 0xb6, 0xc7, 0x8e, 0x8f, 0x92, 0x80,
  0xd4, 0x90, 0xd2, 0xd3, 0xde, 0xd6, 0xd7, 0xd8,
  0xd1, 0xa5, 0xe3, 0xe0, 0xe2, 0xe5, 0x99, 0x9e,
  0x9d, 0xeb, 0xe9, 0xea, 0x9a, 0xed, 0xe7, 0xe1,
  0x85, 0xa0, 0x83, 0xc6, 0x84, 0x86, 0x91, 0x87,
  0x8a, 0x82, 0x88, 0x89, 0x8d, 0xa1, 0x8c, 0x8b,
  0xd0, 0xa4, 0x95, 0xa2, 0x93, 0xe4, 0x94, 0xf6,
  0x9b, 0x97, 0xa3, 0x96, 0x81, 0xec, 0xe8, 0x98,
};

unsigned char tbl_cp2iso[128] = 
{
  0xc7, 0xfc, 0xe9, 0xe2, 0xe4, 0xe0, 0xe5, 0xe7,
  0xea, 0xeb, 0xe8, 0xef, 0xee, 0xec, 0xc4, 0xc5,
  0xc9, 0xe6, 0xc6, 0xf4, 0xf6, 0xf2, 0xfb, 0xf9,
  0xff, 0xd6, 0xdc, 0xf8, 0xa3, 0xd8, 0xd7, 0x80,
  0xe1, 0xed, 0xf3, 0xfa, 0xf1, 0xd1, 0xaa, 0xba,
  0xbf, 0xae, 0xac, 0xbd, 0xbc, 0xa1, 0xab, 0xbb,
  0x81, 0x82, 0x83, 0x84, 0x85, 0xc1, 0xc2, 0xc0,
  0xa9, 0x86, 0x87, 0x88, 0x89, 0xa2, 0xa5, 0x8a,
  0x8b, 0x8c, 0x8d, 0x8e, 0x8f, 0x90, 0xe3, 0xc3,
  0x91, 0x92, 0x93, 0x94, 0x95, 0x96, 0x97, 0xa4,
  0xf0, 0xd0, 0xca, 0xcb, 0xc8, 0x98, 0xcd, 0xce,
  0xcf, 0x99, 0x9a, 0x9b, 0x9c, 0xa6, 0xcc, 0x9d,
  0xd3, 0xdf, 0xd4, 0xd2, 0xf5, 0xd5, 0xb5, 0xde,
  0xfe, 0xda, 0xdb, 0xd9, 0xfd, 0xdd, 0xaf, 0xb4,
  0xad, 0xb1, 0x9e, 0xbe, 0xb6, 0xa7, 0xf7, 0xb8,
  0xb0, 0xa8, 0xb7, 0xb9, 0xb3, 0xb2, 0x9f, 0xa0
};


src/nfsc.h
#define HAVE_NFSSUPPORT
#define HAVE_TERMEMU

#define VERSION_NUM    54		/* must be integer */
#define VERSION_STRING "5.4"

#define PREFIX			0x80		/* additional for the new prot*/

#define XY_XMODEM		1
#define XY_XMODEMCRC		2
#define XY_XMODEMCRC1K		3
#define XY_YMODEM		4
#define XY_YMODEM1K		5
#define XY_YMODEMG		6
#define XY_YMODEMG1K		7

#define ON 			1
#define OFF			0
#define NR_OF_DEVICES		5

#define MAXCOLS 80		/* Longest line possible */

#define P_WRITE(_h, _p, _n) serial_write (_p, _n)
#define P_READ(_h, _p, _n)  { p_read  (_h, _p, (_n)); stats_read  += (_n); }

struct parameters {
	char *name;
	char type;	/* 0 string 
	                   1 numeric (unsigned character)
			   2 function
			   3 offon
			   4 choicelist
	                   5 numeric (unsigned word)
			   */
	unsigned char *value;
	char **arg;
};

struct fkey	/* Function key */
{
	UWORD key;
	UBYTE modifier;
	struct fkey *next;
};

struct sendexpect {
	char *txt;
	struct sendexpect *next;
};

typedef struct
{
  E_MESSAGE m; /* Defined in EPOC.H */
  UWORD arg1; /* First argument */
  UWORD arg2; /* Second argument */
} MESS;

#define QueueTimer(when) p_ioa4(timH, P_FRELATIVE, &timstat, &when)


extern UBYTE  charwidth,charheight, ascent, rows, cols;   
extern WORD   t_w, t_h;
extern UINT   main_win;
extern INT    hgc;
extern BYTE   debug;
extern VOID   *serial;
extern char   cur_appl;		/* Keypad mode or not / vt100 */
extern unsigned char *sb_buf, *sb_nl;			
extern int sb_beg, sb_end;
extern WORD   mstat;
extern HANDLE sendtopid;
extern char   *sendtoaddr;
extern char   mfb[];


extern void   serial_write(char *buf, int len);
extern void   AddToHistory (char ch);
extern void   DoXYmodem (char ch);
extern int    Check (INT val,TEXT *msg);
extern void   Sw_Serial (char);
GLREF_C VOID  TtyEmu(unsigned char *buf, INT len);
GLREF_C VOID  Reset(void);
extern unsigned char iso2cp(unsigned char ch);

extern void   AddFnKey(int, int, char *);
extern void   SendFnKey(struct fkey *);
extern void   InitParams(void);
extern int    LoadParams(char *fname, int warn);
extern int    SaveParams(char *fname, int warn);
extern void   DoExit(int);
extern void   DoPaste(void);
extern void   DoBring(void);
extern void   Conv2Ser(void);
extern void   SetScrollback(int);
extern int    DoCopy(int key, int modifier);
extern void   DoRegister(int state);
extern void   DoSrvBring(void);
extern int    getsbll(int row);
extern void   ParseKey(UWORD key, UBYTE modifier);
extern void   SendModem(void);

extern unsigned char pm_baud, pm_databits, pm_stopbits, pm_parity, pm_xonxoff,
	pm_nortscts, pm_dodsr, pm_dodcd, pm_dodtr, pm_ignpar, pm_statusline,
	pm_clockwin, pm_protocol, pm_jumpscroll, pm_debugterm, pm_font,
	pm_fontsize, pm_isotoibm, pm_maxlines, pm_updfreq, pm_xymodem,
	pm_autodial, pm_port;
extern	WORD	pm_scrollback;
extern char pm_xyname[129];
extern char pm_dialpgm[129];
extern char pm_devices[NR_OF_DEVICES];


src/paste.c
/*
 * Copyright 1993-1996 Rudolf Koenig, Michael Schroeder, Juergen Weigert
 * see the GNU Public License V2 for details
 *
 * In this file lines end with CR/LF, as our psion C-compiler lives in the 
 * DOS-WORLD
 */

/* This module was inspired by the link-paste example from psion */
/* Thanks to Colly for his modules (BTW Colly who?) */ 

#include <plib.h>
#include <p_serial.h>
#include <nfsc.h>
#ifdef HAVE_TERMEMU
#include <hwif.h>
#include "lp.h"

static MESS *pM;
static HANDLE clpid = 0;
static int clwhere;

static char c_nodata[] = "No suitable data";
HANDLE sendtopid = 0;
char  *sendtoaddr= 0;

void
DoBring()
{
  HANDLE srvPid;
  ULONG  fmt, *pfmt;
  UWORD	 args[2];
  int    ret;

  pfmt = &fmt;
  srvPid = p_msendreceivew(p_pidfind("SYS$WSRV.*"), SY_LINK_PASTE, &pfmt);
  if(srvPid <= 0 || !(fmt & DF_LINK_TEXT))
    {
      wInfoMsgCorner(c_nodata, W_CORNER_BOTTOM_RIGHT);
      return;
    }

  args[0] = DF_LINK_TEXT_VAL;
  args[1] = 0;
  if((ret = p_msendreceivew(srvPid, TY_LINKSV_STEP, args)) == 0)
    {
      for(;;)
        {
	  args[0] = (UWORD)mfb;
	  args[1] = 255;
	  if((ret = p_msendreceivew(srvPid, TY_LINKSV_STEP, args)) <= 0)
	    break;
	  mfb[ret++] = '\n';
	  P_WRITE(serial, mfb, ret);
	}
      if (ret == E_FILE_EOF)
        ret = 0;
    }
}

void
DoRegister(int state)
{
  ULONG  fmt;

  if(state == 0)
    {
      clpid = 0;
      Check(p_minit(4, sizeof(MESS)-sizeof(E_MESSAGE)), "Initialise messaging");
      p_mreceive(&mstat, &pM);
    }
  fmt = DF_LINK_TEXT;
  p_msendreceivew(p_pidfind("SYS$WSRV.*"), SY_LINK_SERVER, &fmt);
}

static int
getaline(int where, int *next, int buflen)
{
  int ret, line, start;

  /* Compute the length of the string to copy */
  line = where / MAXCOLS;
  start = line * MAXCOLS;
  ret = getsbll(line) - (where - start);

  line++;
  if(line > pm_scrollback)
    line -= pm_scrollback;
  *next = line * MAXCOLS;

  if(start <= sb_end && start + MAXCOLS > sb_end)
    {
      *next = sb_end;
      if(sb_end - where < ret)
	ret = sb_end - where;
    }
  if(ret > buflen)
    {
      ret = buflen;
      *next = where + ret;
    }
  return ret;
}

void
DoPaste()
{
  int ret, next, where;

  if(sb_beg == -1)
    {
      wInfoMsgCorner(c_nodata, W_CORNER_BOTTOM_RIGHT);
      return;
    }
  where = sb_beg;
  while(where != sb_end)
    {
      ret = getaline(where, &next, 128);
      P_WRITE(serial, (char *)sb_buf+where, ret);
      where = next;
    }
}

void
DoSrvBring(void)
{
  int ret, next;

  ret=E_GEN_NSUP;
  switch(pM->m.type)
    {
      case TY_LINKSV_STEP:
	if(clpid != pM->m.pid)
	  {
	    if(clpid != 0)
	      p_logoff(clpid, TY_LINKSV_DEATH);
	    if(pM->arg1 == DF_LINK_TEXT_VAL)
	      {
		clpid = pM->m.pid;
		p_logon(clpid, TY_LINKSV_DEATH);
		clwhere = sb_beg;
		ret = 0;
	      }
	  }
	else if(clwhere == sb_end || sb_beg == -1 || !pM->arg1 || !pM->arg2)
	  {
	    p_logoff(clpid,TY_LINKSV_DEATH);
	    clpid = 0;
	    ret=E_FILE_EOF;
	  }
	else
	  {
	    /* Compute the length of the string to copy */
	    ret = getaline(clwhere, &next, pM->arg2);
	    p_pcpyto(clpid, (VOID*)pM->arg1, (VOID*)(sb_buf+clwhere), ret);
	    clwhere = next;
	  }
        break;
      case TY_LINKSV_DEATH:
	clpid = 0;
	ret = 0;
        break;

  /* The next 3 types are for OPL support */

      case 0x40: /* opl sent us sumething */

	p_pcpyfr(pM->m.pid, (char *)pM->arg1, mfb, 1);
	p_pcpyfr(pM->m.pid, (char *)pM->arg1+1, mfb+1, *(unsigned char *)mfb);

	/* Disable sending the data, as this would lock up nfsc & the OPL */
	next = sendtopid; sendtopid = 0;
	P_WRITE(serial, (char *)mfb+1, *(unsigned char *)mfb);
	wInfoMsgCorner("Dialstring", W_CORNER_BOTTOM_RIGHT);
	sendtopid = next;
	break;

      case 0x41: /* opl waits for a string */
	sendtopid = pM->m.pid;
	sendtoaddr= (char *)pM->arg1;
	SendModem();
	break;

      case 0x42: /* opl wants to exec menu entry */
	p_mfree(pM,ret);		/* Exit Handling */
	p_mreceive(&mstat,&pM);
	ParseKey(W_SPECIAL_KEY | pM->arg1, 0);
	return;

    }

  p_mfree(pM,ret);
  p_mreceive(&mstat,&pM);
}
#endif


src/params.c
/*
 * Copyright 1993-1996 Rudolf Koenig, Michael Schroeder, Juergen Weigert
 * see the GNU Public License V2 for details
 *
 * In this file lines end with CR/LF, as our psion C-compiler lives in the 
 * DOS-WORLD
 */

#include <plib.h> 
#include <wlib.h> 
#include <p_serial.h>
#include "nfsc.h"

#define DEFAULT_NFSDAEMON

static char c_none[] = "none",
	    c_small[] ="small";

char *portlist[] = {c_none, "TTY:A", "TTY:B", "TTY:I", "TTY:SRX", 0};
char *dev_name[] = {"LOC::M:", "LOC::A:", "LOC::B:", "LOC::C:", "ROM::", 0};
char *baudlist[] = {	  "50",  "75", "110",  "134",  "150",  "300",
			 "600","1200","1800", "2000", "2400", "3600",
			"4800","7200","9600","19200","38400","57600",
			"115200", 0};
char *databitlist[] = {	"5", "6", "7", "8", 0 };
char *paritylist[] = {	c_none, "even", "odd", "mark", "space", 0 };
char *clocklist[] = {	c_none, "big", c_small, 0 };
char *fontlist[] = {	"roman", "swiss", c_small, 0 };
char *fsizelist[] = {	"8", "11", "13", "16", 0 };
char *xylist[] = {	"-", "xmodem", "xmodemcrc", "xmodemcrc1k", "ymodem",
			"ymodem1k", "ymodemg", "ymodemg1k", 0};
char *protolist[] = {	"terminal", "old", "new", 0, };
char cansave = 1;

static int GetDevs(char *, int, int *);
static int DoFnKeys(char *, int, int *);
extern char cant_open_the_file[];

unsigned char
	pm_baud		= P_BAUD_19200-1,
	pm_databits	= P_DATA_8,
	pm_stopbits	= 1,
	pm_port		= 1,
	pm_parity	= 0,
	pm_xonxoff	= 0,
	pm_nortscts	= 0,
	pm_dodsr	= 0,
	pm_dodcd	= 0,
	pm_dodtr	= 1,
	pm_ignpar	= 1,
	pm_statusline	= 0,
	pm_clockwin	= 0,
#ifdef DEFAULT_NFSDAEMON
	pm_protocol	= 2,
#else
	pm_protocol	= 0,
#endif
	pm_jumpscroll	= 1,
	pm_debugterm	= 0,
	pm_font		= 0,
	pm_fontsize	= 0,
	pm_isotoibm	= 1,
	pm_xymodem	= 4,
	pm_maxlines	= 25,
	pm_autodial	= 0,
	pm_updfreq	= 20;
WORD	pm_scrollback	= 100;
char    pm_xyname[129]	= "X\\opd\\xmodem.out";
char    pm_dialpgm[129]	= "X\\opo\\dialme.opo";
char    pm_devices[NR_OF_DEVICES] = { 1, 1, 1, 1, 1};


struct parameters pm[] = {
	{"port",		4,	&pm_port,	portlist},
	{"baudrate",		4,	&pm_baud,	baudlist},
	{"databits",		4,	&pm_databits,	databitlist},
	{"stopbits",		1,	&pm_stopbits,	0},
	{"parity",		4,	&pm_parity,	paritylist},
	{"sendxonxoff",		3,	&pm_xonxoff,	0},
	{"ignorertscts",	3,	&pm_nortscts,	0},
	{"obeydsr",		3,	&pm_dodsr,	0},
	{"obeydcd",		3,	&pm_dodcd,	0},
	{"assertdtr",		3,	&pm_dodtr,	0},
	{"ignoreparity",	3,	&pm_ignpar,	0},
	{"device",		2,	(unsigned char *)GetDevs,	0},
	{"statusline",		3,	&pm_statusline,	0},
	{"clockwindow",		4,	&pm_clockwin,	clocklist},
	{"protocol",		4,	&pm_protocol,	protolist},
	{"jumpscroll",		3,	&pm_jumpscroll,	0},
	{"debugterm",		3,	&pm_debugterm,	0},
	{"font",		4,	&pm_font,	fontlist},
	{"fontsize",		4,	&pm_fontsize,	fsizelist},
	{"xymodem",		4,	&pm_xymodem,	xylist},
	{"isotoibm",		3,	&pm_isotoibm,	0},
	{"maxlines",		1,	&pm_maxlines,	0},
	{"autodial",		3,	&pm_autodial,	0},
	{"fnkey",		2,	(unsigned char *)DoFnKeys,	0},
	{"dialpgm",		0,	(unsigned char *)pm_dialpgm+1,	0},
	{"xymodemout",		0,	(unsigned char *)pm_xyname+1,	0},
	{"scrollback",		5,	(unsigned char *)&pm_scrollback,0},
	{0,			0,	0,		0}
};

static int 
GetDevs(buf, load, which)
  char *buf;
  int load, *which;
{
  if(load)
    {
      if(*which == 0)
	for(load = 0;dev_name[load]; load++)
	  pm_devices[load] = 0;

      (*which)++;
      for(load = 0;dev_name[load]; load++)
        if(p_scmpi(dev_name[load], buf) == 0)
	  {
	    pm_devices[load] = 1;
	    return 0;
	  }
      return 1;
    }
  else
    {
      for(load = *which; dev_name[load] != 0; load++)
        if(pm_devices[load])
	  {
	    p_scpy(buf, dev_name[load]);
	    *which = load+1;
	    return 1;
	  }
      return 0;
    }
}

static int 
DoFnKeys(buf, load, which)
  char *buf;
  int load, *which;
{
#ifdef HAVE_TERMEMU
  extern struct fkey *fk;
  static struct fkey *f = (struct fkey *)-1;
  UWORD key, modifier;
  char *p;
  int len;

  if(load)
    {
      len = p_slen(buf);
      p = buf;   p_stog(&p, &key, 16);
      p = buf+5; p_stog(&p, &modifier, 16);
      p = buf+7; *p = len - 8;
      AddFnKey(key, modifier, p);
      return 0;
    }
  else
    {
      if(f == (struct fkey *)-1)
        f = fk;
      if(f == 0)
        {
	  f = (struct fkey *)-1;
	  return 0;
	}
      p_atos(buf, "%04x %02x %s", f->key, f->modifier, (char *)f+sizeof(*f));
      f = f->next;
      return 1;
    }
#endif
  return 0;
}

#define BLEN 256
void
InitParams()
{
  extern ULONG timint;

  timint = pm_updfreq;
  if(pm_font == 0) pm_fontsize += 4;
  if(pm_font == 1) pm_fontsize += 8;
  if(pm_font == 2) pm_fontsize = 12;
  *pm_xyname = p_slen(pm_xyname+1);
  *pm_dialpgm = p_slen(pm_dialpgm+1);
#ifdef HAVE_TERMEMU
  SetScrollback(pm_scrollback);
#endif
  Conv2Ser();
}
/*
 * Note: LoadParams and SaveParams require an OPL string
 */
int
LoadParams(fname, warn)
  char *fname;
  int warn;
{
  VOID *fcb;
  WORD num;
  char *p, *p2;
  char **ps;
  int len, i, j, line, which = 0;

  fname[*fname+1] = 0;
  if(p_open(&fcb, fname+1, P_FTEXT | P_FSHARE))
    {
      if(warn)
	wInfoMsgCorner(cant_open_the_file, W_CORNER_BOTTOM_RIGHT);
      return 1;
    }
  line = 0;
  while((len = p_read(fcb, mfb, BLEN-1)) > 0)
    {
      line++;
      if(*mfb == '#' || *mfb == 0)	/* Allow for comments */
        continue;

      /* Chop the line into parameter-name and argument */
      mfb[len] = 0;
      while(mfb[len-1] == '\n' || mfb[len-1] == '\r')
        mfb[--len] = 0;
      p = p_skipch(mfb);
      if(*p) *p++ = 0;
      if(*p) p = p_skipwh(p);
      
      if(p_scmpi(mfb, "include") == 0)
        {
	  cansave = 0;
	  *(p-1) = p_slen(p);
	  if(LoadParams(p-1, warn))
	    return 1;
	  continue;
	}

      for(i = 0; pm[i].name; i++)
        if(p_scmpi(pm[i].name, mfb) == 0)
	  break;
      if(pm[i].name == 0 || *p == 0)
	break;
      
      switch(pm[i].type)
        {
	  case 0:
	    p_scpy((char *)pm[i].value, p);
	    break;
	  case 1:
	    p2 = p;
	    p_stoi(&p2, &num);
	    *pm[i].value = num;
	    break;
	  case 2:
	    if(((int (*)(char *, int, int *))pm[i].value)(p, 1, &which))
	      len = -1;
	    break;
	  case 3:
	    if(!p_scmpi(p, "on") || !p_scmpi(p, "yes"))
	      *pm[i].value = 1;
	    else
	      *pm[i].value = 0;
	    break;
	  case 4:
	    ps = (char **)pm[i].arg;
	    for(j = 0; *ps[j] != 0; j++)
	      if(!p_scmpi(ps[j], p))
	        {
		  *pm[i].value = j;
		  break;
		}
	    if(*ps[j] == 0)
	      len = -1;
	    break;
	  case 5:
	    p2 = p;
	    p_stoi(&p2, (WORD *)pm[i].value);
	    break;
	}
      if(len < 0)
        {
	  *p = 0;
	  break;
	}
    }
  p_close(fcb);

  if(warn)
    {
      if(pm[i].name == 0 || *p == 0)
	{
	  p_atos(mfb, "Bogus %s in line %d", *p == 0 ? "value" : "name", line);
	  wInfoMsgCorner(mfb, W_CORNER_BOTTOM_RIGHT);
	  p_sleep(20);
	  return 1;
	}
      else
	wInfoMsgCorner("Loaded", W_CORNER_BOTTOM_RIGHT);
    }

  InitParams();
  return 0;
}

int
SaveParams(fname, warn)
  char *fname;
  int warn;
{
  VOID *fcb;
  char **ps, *p;
  int i, which = 0;

  if(pm_fontsize == 12) 
    pm_font = 2, pm_fontsize = 0;
  else if(pm_fontsize >= 8)
    pm_font = 1, pm_fontsize -= 8;
  else if(pm_fontsize >= 4)
    pm_font = 0, pm_fontsize -= 4;
  else
    pm_font = 0, pm_fontsize = 0;

  fname[*fname+1] = 0;
  if(p_open(&fcb, fname+1, P_FTEXT | P_FREPLACE | P_FUPDATE))
    {
      if(warn)
	wInfoMsgCorner(cant_open_the_file, W_CORNER_BOTTOM_RIGHT);
      return;
    }
  p_atos(mfb, "# Do not append spaces at the end of line");
  p_write(fcb, mfb, p_slen(mfb));
  for(i = 0; pm[i].name; i++)
    {
      p_atos(mfb, "%s\t%s", pm[i].name, p_slen(pm[i].name) < 8 ? "\t" : "");
      p = mfb + p_slen(mfb);
      switch(pm[i].type)
        {
	  case 0:
      	    p_scpy(p, (char *)pm[i].value);
	    break;
	  case 1:
      	    p_atos(p, "%d", (WORD)(*pm[i].value));
	    break;
	  case 2:
	    if(((int (*)(char *, int, int *))pm[i].value)(p, 0, &which))
	      i--;
	    else
	      continue;
	    break;
	  case 3:
	    p_scpy(p, *pm[i].value ? "on" : "off");
	    break;
	  case 4:
	    ps = (char **)pm[i].arg;
	    p_scpy(p, ps[*pm[i].value]);
	    break;
	  case 5:
      	    p_atos(p, "%d", *(WORD *)pm[i].value);
	    break;
	}
      p_write(fcb, mfb, p_slen(mfb));
    }
  p_close(fcb);
  if(warn)
    wInfoMsgCorner("Saved", W_CORNER_BOTTOM_RIGHT);
  return 0;
}
